pkg/v1/layout/write.go

```diff
@@ -0,0 +1,318 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package layout
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/empty"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+	"golang.org/x/sync/errgroup"
+)
+
+var layoutFile = `{
+    "imageLayoutVersion": "1.0.0"
+}`
+
+// AppendImage writes a v1.Image to an OCI image layout at path and updates
+// the index.json to reference it.
+func AppendImage(path string, img v1.Image, options ...LayoutOption) (v1.ImageIndex, error) {
+	if err := WriteImage(path, img); err != nil {
+		return nil, err
+	}
+
+	mt, err := img.MediaType()
+	if err != nil {
+		return nil, err
+	}
+
+	d, err := img.Digest()
+	if err != nil {
+		return nil, err
+	}
+
+	manifest, err := img.RawManifest()
+	if err != nil {
+		return nil, err
+	}
+
+	desc := v1.Descriptor{
+		MediaType: mt,
+		Size:      int64(len(manifest)),
+		Digest:    d,
+	}
+
+	for _, opt := range options {
+		if err := opt(&desc); err != nil {
+			return nil, err
+		}
+	}
+
+	return AppendDescriptor(path, desc)
+}
+
+// AppendIndex writes a v1.ImageIndex to an OCI image layout at path and updates
+// the index.json to reference it.
+func AppendIndex(path string, ii v1.ImageIndex, options ...LayoutOption) (v1.ImageIndex, error) {
+	if err := WriteIndex(path, ii); err != nil {
+		return nil, err
+	}
+
+	mt, err := ii.MediaType()
+	if err != nil {
+		return nil, err
+	}
+
+	d, err := ii.Digest()
+	if err != nil {
+		return nil, err
+	}
+
+	manifest, err := ii.RawIndexManifest()
+	if err != nil {
+		return nil, err
+	}
+
+	desc := v1.Descriptor{
+		MediaType: mt,
+		Size:      int64(len(manifest)),
+		Digest:    d,
+	}
+
+	for _, opt := range options {
+		if err := opt(&desc); err != nil {
+			return nil, err
+		}
+	}
+
+	return AppendDescriptor(path, desc)
+}
+
+// AppendDescriptor adds a descriptor to the index.json of an ImageIndex located
+// at path.
+func AppendDescriptor(path string, desc v1.Descriptor) (v1.ImageIndex, error) {
```

Comment:
I'd like to understand the relationship between the image layout and the returned `v1.ImageIndex` on this and other similar functions. Reading the [spec](https://github.com/opencontainers/image-spec/blob/master/image-layout.md) didn't help particularly because it seems that image indices can refer to other image indices. So is the intention that the returned image index always represents the whole enchilada or for some functions, does it represent one of the constituent image indices?

Reply:
I'll attempt to explain but let me know if I can make this clearer in the comments or in general:

An [image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md) is just a JSON data structure (AKA a manifest) that can refer to other manifests. Those manifests *usually* represent either an image or another index.

An [image layout](https://github.com/opencontainers/image-spec/blob/master/image-layout.md) is a way to store images on disk comprised of a "blobs/" directory with content-addressable contents and an "index.json" that is an index those contents and acts as an entrypoint. That index.json file's format is an [image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md).

These top-level functions interact with either the "blobs/" directory or the "index.json" or both.

`WriteBlob`, `WriteImage`, and `WriteIndex` add files to the "blobs/" directory, but don't update index.json. This is useful for doing batches or writes without having to update the index.json in-betweeen.

`AppendImage` and `AppendIndex` add files to the "blobs" directory *and* update the index.json. I expect these would be the most common write operations.

`Write` takes an ImageIndex and essentially converts it to an image layout, where the image index manifest becomes the top-level index.json, and everything it references is written to the "blobs/" directory. This is useful for serializing an index to disk without having an extraneous wrapper around it.

The `layout.Image` and `layout.Index` functions for reading images/indexes mirrors what we're doing with `remote.Image` and `remote.Index` -- they just read from the data source and produce an image or index.

Reply:
That's clear. I guess a clarification which would help me is to document the return values of the public functions. For example, `AppendDescriptor` might document its return values as follows:
```
// AppendDescriptor adds a descriptor to the index.json of an ImageIndex located
// at path. If successful, it returns a nil error and an ImageIndex representing the updated
// index.json (and equivalent to that which would be returned by calling Index(path) after 
// AppendDescriptor).
```

Reply:
Alternatively, you might wish to remove the `ImageIndex` return values. The API provides the `Index` function to get hold of an `ImageIndex` if necessary.